src/Services/TableService.php

```diff
@@ -47,39 +52,167 @@ public function findMany(int|string|array $id, string $fieldName = 'Id'): Collec
+    {
+        try {
+            Forrest::sobjects(implode('/', [
+                $this->table,
+                $id,
+            ]), [
+                'method' => 'patch',
+                'body' => $data,
+            ]);
+
+            return $this->find($id);
+        } catch (SalesforceException $ex) {
+            $this->throwException($ex);
             throw $ex;
         }
     }
 
+    /**
+     * Upsert a record using the specified external Id field and Id with the provided data
+     *
+     * @param array<string, mixed> $data
+     * @return array<string, mixed>|false
+     */
+    public function createOrUpdate(string $field, string $id, array $data = []): array|false
+    {
+        try {
+            $response = Forrest::sobjects(implode('/', [
+                $this->table,
+                $field,
+                $id,
+            ]), [
+                'method' => 'patch',
+                'body' => $data,
+            ]);
+
+            if ($response['success'] ?? false) {
+                $response['data'] = $this->find($response['id']);
+                return $response;
+            }
+        } catch (SalesforceException $ex) {
+            $this->throwException($ex);
+            throw $ex;
+        }
+
+        return false;
+    }
+
+    /**
+     * Delete a record on the table using the specified Id
+     *
+     * @return bool
+     */
+    public function delete(string $id): bool
+    {
+        try {
+            Forrest::sobjects(implode('/', [
+                $this->table,
+                $id,
+            ]), [
+                'method' => 'delete',
+            ]);
+            return true;
+        } catch (SalesforceException $ex) {
+            $this->throwException($ex);
+            throw $ex;
+        }
+    }
+
+    protected function throwException(SalesforceException $ex): void
+    {
+        $error = $this->decodeError($ex);
+
+        if (property_exists($error, 'errorCode')) {
+            $message = $this->formatErrorMessage($error->message);
+
+            switch ($error->errorCode) {
+                case 'ENTITY_IS_DELETED':
+                    throw new EntityIsDeletedException($message);
+                case 'INVALID_CROSS_REFERENCE_KEY':
+                    throw new InvalidCrossReferenceKeyException($message);
+                case 'INVALID_ID_FIELD':
+                    throw new InvalidFieldException($message);
+                case 'INVALID_QUERY_FILTER_OPERATOR':
+                    throw new InvalidFieldException($message);
+                case 'REQUIRED_FIELD_MISSING':
+                    throw new RequiredFieldMissingException($message);
+                case 'MALFORMED_QUERY':
+                    throw new MalformedQueryException($message);
+                case 'MALFORMED_ID':
+                    throw new MalformedIdException($message);
+                case 'NOT_FOUND':
+                    throw new NotFoundException($message);
+            }
+        }
+    }
```

Comment:
would it be better to throw the original exception here and then not throw it manually errywhere else?

```suggestion
        }
        
        throw $ex;
    }
```